test/infrastructure/vcsim/controllers/vmbootstrap.go

```diff
@@ -0,0 +1,1002 @@
+						},
+					},
+				},
+			},
+		},
+	}
+	if err := inmemoryClient.Get(ctx, client.ObjectKeyFromObject(kubeProxyDaemonSet), kubeProxyDaemonSet); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to get kube-proxy DaemonSet")
+		}
+
+		if err := inmemoryClient.Create(ctx, kubeProxyDaemonSet); err != nil && !apierrors.IsAlreadyExists(err) {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to create kube-proxy DaemonSet")
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *vmBootstrapReconciler) reconcileBoostrapCoredns(ctx context.Context, cluster *clusterv1.Cluster, machine *clusterv1.Machine, _ *infrav1.VSphereVM) (ctrl.Result, error) {
+	// No-op if the machine is not a control plane machine.
+	if !util.IsControlPlaneMachine(machine) {
+		return ctrl.Result{}, nil
+	}
+
+	// TODO: Add provisioning time for CoreDNS.
+
+	// Compute the resource group unique name.
+	// NOTE: We are using reconcilerGroup also as a name for the listener for sake of simplicity.
+	resourceGroup := klog.KObj(cluster).String()
+	inmemoryClient := r.InMemoryManager.GetResourceGroup(resourceGroup).GetClient()
+
+	// Create the coredns configMap.
+	corednsConfigMap := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: metav1.NamespaceSystem,
+			Name:      "coredns",
+		},
+		Data: map[string]string{
+			"Corefile": "ANG",
+		},
+	}
+	if err := inmemoryClient.Get(ctx, client.ObjectKeyFromObject(corednsConfigMap), corednsConfigMap); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to get coreDNS configMap")
+		}
+
+		if err := inmemoryClient.Create(ctx, corednsConfigMap); err != nil && !apierrors.IsAlreadyExists(err) {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to create coreDNS configMap")
+		}
+	}
+	// Create the coredns deployment.
+	corednsDeployment := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: metav1.NamespaceSystem,
+			Name:      "coredns",
+		},
+		Spec: appsv1.DeploymentSpec{
+			Template: corev1.PodTemplateSpec{
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:  "coredns",
+							Image: "registry.k8s.io/coredns/coredns:v1.10.1",
+						},
+					},
+				},
+			},
+		},
+	}
+
+	if err := inmemoryClient.Get(ctx, client.ObjectKeyFromObject(corednsDeployment), corednsDeployment); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to get coreDNS deployment")
+		}
+
+		if err := inmemoryClient.Create(ctx, corednsDeployment); err != nil && !apierrors.IsAlreadyExists(err) {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to create coreDNS deployment")
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *vmBootstrapReconciler) reconcileShutdown(ctx context.Context, cluster *clusterv1.Cluster, machine *clusterv1.Machine, conditionsTracker *infrav1.VSphereVM) (ctrl.Result, error) {
```

Comment:
We are we calling this here all shutdown while it's delete in core CAPI. Is there a semantic difference?